File: barq-sdk-ts/test/options_client.test.js
const test = require("node:test");
const assert = require("node:assert/strict");
const { BarqClient, GrpcClient } = require("../dist/index.js");

test("typescript grpc client encodes insert options", async () => {
    const client = new GrpcClient("127.0.0.1:50051");
    let seenRequest;

    client.client = {
        insert(request, metadata, callback) {
            seenRequest = request;
            callback(null, {});
        },
    };

    await client.insert("docs", "ts-doc", [1.0, 0.0], { sdk: "ts" }, { waitForCommit: false });

    assert.equal(seenRequest.options.waitForCommit, false);
});

test("typescript grpc client encodes search options with default fallback", async () => {
    const client = new GrpcClient("127.0.0.1:50051");
    let seenRequest;

    client.client = {
        search(request, metadata, callback) {
            seenRequest = request;
            callback(null, { results: [] });
        },
    };

    await client.search("docs", [1.0, 0.0], 1, { consistency: "followers" });

    assert.equal(seenRequest.options.consistency, "CONSISTENCY_FOLLOWERS");
    assert.equal(seenRequest.options.allowFallback, true);
});

test("typescript public client rejects options on rest fallback", async () => {
    const client = new BarqClient({
        baseUrl: "http://127.0.0.1:8080",
        apiKey: "test-key",
    });

    await assert.rejects(
        () => client.collection("docs").search([1.0, 0.0], "hello", 1, undefined, { allowFallback: false }),
        /advanced search options are only supported for vector-only gRPC search/,
    );
});

File: barq-sdk-go/options_client_test.go
package barq

import (
	"context"
	"strings"
	"testing"

	pb "github.com/YASSERRMD/barq-db/barq-sdk-go/proto"
)

func TestProtoInsertOptionsEncodeWaitForCommit(t *testing.T) {
	waitForCommit := false

	options := protoInsertOptions(&InsertOptions{WaitForCommit: &waitForCommit})
	if options == nil {
		t.Fatal("expected insert options to be encoded")
	}
	if options.WaitForCommit {
		t.Fatal("expected wait_for_commit to remain false")
	}
}

func TestProtoSearchOptionsDefaultAllowFallback(t *testing.T) {
	consistency := ConsistencyFollowers

	options := protoSearchOptions(&SearchOptions{Consistency: &consistency})
	if options == nil {
		t.Fatal("expected search options to be encoded")
	}
	if options.Consistency != pb.Consistency_CONSISTENCY_FOLLOWERS {
		t.Fatalf("unexpected consistency %v", options.Consistency)
	}
	if !options.AllowFallback {
		t.Fatal("expected allow_fallback default to remain true")
	}
}

func TestClientSearchRejectsAdvancedOptionsOnRestFallback(t *testing.T) {
	allowFallback := false
	client := NewClient(Config{BaseURL: "http://127.0.0.1:8080", APIKey: "test-key"})

	_, err := client.Search(context.Background(), "docs", SearchRequest{
		Vector:  []float32{1.0, 0.0},
		Query:   "hello",
		TopK:    1,
		Options: &SearchOptions{AllowFallback: &allowFallback},
	})
	if err == nil {
		t.Fatal("expected search to reject advanced options on REST fallback")
	}
	if !strings.Contains(err.Error(), "advanced search options are only supported for vector-only gRPC search") {
		t.Fatalf("unexpected error: %v", err)
	}
}
